pymc_marketing/customer_choice/nested_logit.py

```diff
@@ -0,0 +1,783 @@
+            self.nest_indices = {"top": top_level, "mid": mid_level}
+        else:
+            all_nests = list(top_level.keys())
+            self.nest_indices = {"top": top_level}
+
+        self.all_nests = all_nests
+        self.lambda_lkup = dict(zip(all_nests, range(len(all_nests)), strict=False))
+
+        # note: type hints for coords required for mypy to not get confused
+        self.coords: dict[str, list[str]] = self._prepare_coords(
+            choice_df,
+            self.alternatives,
+            self.covariates,
+            self.fixed_covar,
+            self.all_nests,
+        )
+
+        return X, F, y
+
+    def build_model(self, X, y, **kwargs):
+        """Do not use, required by parent class. Prefer make_model()."""
+        return super().build_model(X, y, **kwargs)
+
+    def make_exp_nest(self, U, W, betas_fixed_, lambdas_nests, nest, level="top"):
+        """Calculate within nest probabilities.
+
+        Allows mixing between levels of the nesting structure within the PyMC model.
+        """
+        nest_indices = self.nest_indices
+        lambda_lkup = self.lambda_lkup
+        N = U.shape[0]
+        if "_" in nest:
+            parent, child = nest.split("_")
+        else:
+            parent = None
+        y_nest = U[:, nest_indices[level][nest]]
+        if W is None:
+            w_nest = pm.math.zeros((N, len(self.alternatives)))
+        else:
+            betas_fixed_temp = betas_fixed_[nest_indices[level][nest], :]
+            betas_fixed_temp = pm.Deterministic(
+                f"beta_fixed_{level}_{nest}",
+                pt.set_subtensor(betas_fixed_temp[-1, :], 0),
+            )
+            w_nest = pm.math.dot(W, betas_fixed_temp.T)
+
+        if len(nest_indices[level][nest]) > 1:
+            max_y_nest = pm.math.max(y_nest, axis=0)
+            P_y_given_nest = pm.Deterministic(
+                f"p_y_given_{nest}",
+                pm.math.softmax(y_nest / lambdas_nests[lambda_lkup[nest]], axis=1),
+            )
+        else:
+            max_y_nest = pm.math.max(y_nest)
+            ones = pm.math.ones((N, 1))
+            P_y_given_nest = pm.Deterministic(f"p_y_given_{nest}", ones)
+        if parent is None:
+            lambda_ = lambdas_nests[lambda_lkup[nest]]
+            I_nest = pm.Deterministic(
+                f"I_{nest}", pm.math.logsumexp((y_nest - max_y_nest) / lambda_)
+            )
+            W_nest = w_nest + I_nest * lambda_
+        else:
+            l1 = lambdas_nests[lambda_lkup[nest]]
+            l2 = lambdas_nests[lambda_lkup[parent]]
+            lambdas_ = l1 * l2
+            I_nest = pm.Deterministic(
+                f"I_{nest}", pm.math.logsumexp((y_nest - max_y_nest) / lambdas_)
+            )
+            W_nest = w_nest + I_nest * (lambdas_)
+
+        exp_W_nest = pm.math.exp(W_nest)
+        return exp_W_nest, P_y_given_nest
+
+    def make_P_nest(self, U, W, betas_fixed_, lambdas_nests, level):
```

Comment:
type hints missing and parameter descriptions

Reply:
maybe even add some comments on the code to explain the logic a bit? (It is not super easy to follow for a new person reading for the first time)

Reply:
added